README.md
service.pushbullet.com
======================

Kodi (XBMC) service for receiving push notification via Pushbullet

Work in progress - currently unreleased.

Right now it allows pushing video links and having them play in XBMC.


changelog.txt
Changelog for service.pushbullet.com

*** 0.1.12 ***

Change from using requests2-2.2.1 module to requests-2.3.0
Bump up OAuth module version requirement to 0.0.7

*** 0.1.11 ***

Bump up OAuth module version requirement to 0.0.6

*** 0.1.10 ***

Fix for service dying/not reconnecting when returning from sleep/suspend/hibernate
Generally better handling of disconnects and connection errors

*** 0.1.9 ***

Fix for service being started a second time instead of starting the interface when started from the home screen

*** 0.1.8 ***

Fix the hardcoded fake email I had in there while I was taking a screenshot

*** 0.1.7 ***

Fix for actions operating on the wrong push
Fix for deleting the top push causing focus to drop to the bottom of the list

*** 0.1.6 ***

Localize strings
Added some code to keep service alive through push receiver errors/disconnects

*** 0.1.5 ***

Added ability to switch between viewing pushes for this device, another device or all devices
Fix for note view not showing any text on some skins
Added thumbnails for images (and videos when 'Load Video Thumbnails' is enabled)
Added custom fonts for Aeon MQ 5 and Amber
Added setting 'Force Specific Device Pushes To View On Start'
Added setting 'Default Pushes To View On Start' with the options 'This Device', 'All Devices', and 'Last Selection'
Restructured the settings categories

*** 0.1.4 ***

Fix for some protocols being handled as media file links in some instances
Fix for service not starting on initial install
Fix for service errors when addon is authorized but the device is not linked/added yet

*** 0.1.3 ***

Added crappy fanart :)
Added support for sop:// and acestream:// protocol links (requires P2P-Streams addon)
Addes support for mms://, rtsp:// and rtmp:// protocol links

*** 0.1.2 ***

Added ability to download video page links via youtube-dl module
Added success/fail dialogs on downloads

*** 0.1.1 ***

Added the ability to download file and link pushes
Added the ability to play link and file pushes that point to media files
Added media type sub-icons
List view will now update checkboxes dynamically when toggled remotely
Now caches video thumbnails when 'Load Video Thumbnails' is enabled

*** 0.1.0 ***

Initial public release

main.py
# -*- coding: utf-8 -*-
import sys
from lib import util
T = util.T

def showError(msg):
	import xbmcgui
	xbmcgui.Dialog().ok(T(32051).upper(), '{0}:'.format(T(32051)),'',msg)
	
def loadTokenFromFile():
	import OAuthHelper
	
	token = OAuthHelper.getToken('service.pushbullet.com',from_file=True)
	if token: saveToken(token)

def deviceNameExists(client,name):
	from lib import PushbulletTargets
	try:
		for d in client.getDevicesList():
			if not d.get('active'): continue
			if name == d.get('nickname'): return True
		return False
	except PushbulletTargets.PushbulletException, e:
		showError(e.message)
		return None
	
def addNewDevice(client,device):
	import xbmcgui
	from lib import PushbulletTargets
	
	#TODO: Check for existing device nickname and handle	
	
	device.name = xbmcgui.Dialog().input('{0}:'.format(T(32065)),device.name or T(32066)) or device.name
	if not device.name: return False
	
	while deviceNameExists(client,device.name):
		device.name = xbmcgui.Dialog().input(T(32067),device.name or '')
		if not device.name: return
		
	try:
		client.addDevice(device)
	except PushbulletTargets.PushbulletException, e:
		util.LOG('FAILED TO ADD DEVICE: {0}'.format(device.name))
		showError(e.message)
		return False
	return True
	
def linkDevice():
	import xbmcgui
	from lib import PushbulletTargets
	from lib import devices
	
	token = util.getSetting('token')
	
	if not token:
		xbmcgui.Dialog().ok(T(32068), T(32069),T(32070))
		return
		
	client = PushbulletTargets.Client(token)
	deviceMap = {}
	try:
		for d in client.getDevicesList():
			if not d.get('active'): continue
			deviceMap[d['iden']] = d['nickname']
	except PushbulletTargets.PushbulletException, e:
		showError(e.message)
		return
	
	idx = xbmcgui.Dialog().select(T(32071),deviceMap.values() + [T(32072)])
	if idx < 0: return
	
	dev = devices.KodiDevice(None,util.getSetting('device_name') or None)
	if idx == len(deviceMap):
		if not addNewDevice(client,dev): return
	else:
		dev.ID = deviceMap.keys()[idx]
		dev.name = deviceMap.values()[idx] or dev.name
		
	util.setSetting('device_iden',dev.ID)
	util.setSetting('device_name',dev.name)
	util.LOG('DEVICE LINKED: {0}'.format(dev.name))
	xbmcgui.Dialog().ok(T(32073), '{0}: '.format(T(32074)), '  [B]{0}[/B]'.format(dev.name), T(32075))

def renameDevice():
	import xbmcgui
	from lib import devices
	from lib import PushbulletTargets
	from lib import util
	
	dev = devices.getDefaultKodiDevice(util.getSetting('device_iden'),util.getSetting('device_name'))
	if not dev or not dev.ID:
		xbmcgui.Dialog().ok(T(32076),T(32077))
	name = xbmcgui.Dialog().input('{0}:'.format(T(32078)),dev.name or '')
	if not name: return
	if name == dev.name: return
	
	token = util.getSetting('token')
	
	client = PushbulletTargets.Client(token)
	
	while deviceNameExists(client,name):
		name = xbmcgui.Dialog().input(T(32079),dev.name or '')
		if not name: return
		if name == dev.name: return

	if not token:
		xbmcgui.Dialog().ok(T(32068),T(32080),T(32081))
		return
		
	try:
		if client.updateDevice(dev,nickname=name):
			util.setSetting('device_name',dev.name)
			xbmcgui.Dialog().ok(T(32062),'{0}: '.format(T(32082)),'',dev.name)
			
	except PushbulletTargets.PushbulletException, e:
		showError(e.message)

def selectDevice():
	from lib import gui, PushbulletTargets

	token = util.getSetting('token')

	client = PushbulletTargets.Client(token)

	ID = gui.selectDevice(client,extra=T(32083))
	if ID == None: return
	util.setSetting('selected_device',ID)

def saveToken(token):
	util.setSetting('token',token)

def authorize():
	import OAuthHelper
	
	token = OAuthHelper.getToken('service.pushbullet.com')
	if token: saveToken(token)

def main():
	if not util.getToken():
		import xbmcgui
		xbmcgui.Dialog().ok(T(32084),T(32085),T(32086))
		util.ADDON.openSettings()
		return
		
	if not util.getSetting('device_iden'):
		import xbmcgui
		xbmcgui.Dialog().ok(T(32084),T(32087),T(32088),T(32089))
		util.ADDON.openSettings()
		return

	from lib import gui
	gui.start()

def handleArg(arg):
	if arg == 'LINK_DEVICE':
		linkDevice()
	elif arg == 'RENAME_DEVICE':
		renameDevice()
	elif arg == 'TOKEN_FROM_FILE':
		loadTokenFromFile()
	elif arg == 'SELECT_DEVICE':
		selectDevice()
	elif arg == 'AUTHORIZE':
		authorize()

if __name__ == '__main__':
	try:
		args = None
		if len(sys.argv) > 1:
			args = sys.argv[1:]
		
		if args:
			handleArg(args[0])
		else:
			main()
	except:
		import traceback
		traceback.print_exc()

service.py
# -*- coding: utf-8 -*-
import sys, time
import xbmc
from lib import util, pushhandler
from lib import PushbulletTargets
PushbulletTargets.LOG = util.LOG

from lib import devices
import YDStreamUtils as StreamUtils

def ERROR():
	import traceback
	xbmc.log(traceback.format_exc())

class TargetsBox(object):
	def __init__(self):
		self.token = util.getSetting('token')
		self.deviceID = util.getSetting('device_iden')
		self.deviceName = util.getSetting('device_name')
		self.device = None
		self.targets = None

		self.reconnectDelay = 5
		self.nextReconnect = 0
		self.hasConnected = False
		self.badToken = False
		self.start()
		self.connect()

	def start(self):
		self.targets = PushbulletTargets.Targets(
			self.token,
			most_recent=util.getSetting('most_recent',0),
			most_recent_callback=self.mostRecentUpdated
		)

		self.device = devices.getDefaultKodiDevice(self.deviceID,self.deviceName)
		self.targets.registerDevice(self.device)

	def ready(self):
		if self.badToken: return False
		if self.hasConnected:
			return not self.badToken and not self.targets.terminated and self.readyToReconnect()
		else:
			if not self.readyToReconnect(): return False
			self.connect()

	def join(self,timeout=0.2):
		self.targets._th.join(timeout=timeout)

	def mostRecentUpdated(self,modified):
		util.setSetting('most_recent','{0:10f}'.format(modified))

	def close(self):
		try:
			self.targets.close(force=True)
			self.targets._th.join()
		except RuntimeError:
			pass

	def updateReconnectDelay(self):
		if self.reconnectDelay > 60: return
		self.reconnectDelay *= 2

	def readyToReconnect(self):
		return time.time() > self.nextReconnect

	def connect(self):
		try:
			self.targets.connect()
			self.hasConnected = True
			self.badToken = False
			self.reconnectDelay = 5
		except PushbulletTargets.ws4py.exc.HandshakeError, e:
			self.badToken = True
			if '401' in e.msg:
				util.LOG('CONNECT HANDSHAKE ERROR - BAD TOKEN')
			else:
				ERROR('CONNECT HANDSHAKE ERROR')
			return
		except:
			ERROR()
			self.nextReconnect = time.time() + self.reconnectDelay
			util.LOG('CONNECT ERROR: Waiting {0} seconds'.format(self.reconnectDelay))
			self.updateReconnectDelay()
			return

class PushbulletService(xbmc.Monitor):
	def __init__(self):
		self.targetsBox = None
		self.start()

	def onAbortRequested(self):
		if not self.targetsBox: return
		self.targetsBox.close()

	def onSettingsChanged(self):
		self.loadSettings()

	def credentialsUpdated(self):
		token = util.getSetting('token')
		deviceID = util.getSetting('device_iden')
		if not self.targetsBox:
			if token and deviceID:
				return True
			else:
				return None
		if token != self.targetsBox.token:
			util.LOG('TOKEN CHANGED')
			return True

		if deviceID != self.targetsBox.deviceID:
			util.LOG('DEVICE CHANGED FROM: {0} TO: {1}'.format(self.targetsBox.deviceName,util.getSetting('device_name')))
			return True

		return False
		
	def loadSettings(self):
		self.reconnectDelay = 5
		self.interruptMedia = util.getSetting('interrupt_media',False)
		self.instantPlay = util.getSetting('instant_play',True)
		self.showNotification = util.getSetting('show_notification',True)
		
		if self.credentialsUpdated():
			if self.targetsBox: self.targetsBox.close()
			self.targetsBox = TargetsBox()

	def start(self):
		util.LOG('SERVICE: STARTED {0}'.format(util.ADDON.getAddonInfo('version')))
		
		self.loadSettings()
		
		while not xbmc.abortRequested:
			if self.targetsBox and self.targetsBox.ready():
				self.active()
			else:
				self.idle()
		self.done()
		
	def runServer(self):
		while self.targetsBox.ready() and not xbmc.abortRequested:
			self.targetsBox.join()
			if self.targetsBox.device.hasPush():
				if self.instantPlay:
					if not StreamUtils.isPlaying() or self.interruptMedia:
						data = self.targetsBox.device.getNext()
						if data: pushhandler.handlePush(data)
				else:
					if self.showNotification:
						data = self.targetsBox.device.getNext()
						if data:
							util.notify(
								'{0}: {1}'.format(util.T(32090),data.get('type','?')),
								data.get('title','')
							)
					self.targetsBox.device.clear()

			xbmc.sleep(200)
		
		self.targetsBox.close()
		self.targetsBox = None

	def active(self):
		util.LOG('SERVICE: ACTIVE')
				
		try:
			self.runServer()
		except:
			util.ERROR('Server Error')
			util.notify(util.T(32105),util.T(32106))
			xbmc.sleep(5000)

		if not xbmc.abortRequested: #If not shutting down, reset in case we died of error
			self.loadSettings()
		
	def idle(self):
		util.LOG('SERVICE: IDLING')
		while not (self.targetsBox and self.targetsBox.ready()) and not xbmc.abortRequested:
			xbmc.sleep(1000)

	def done(self):
		util.LOG('SERVICE: DONE')

if __name__ == '__main__':
	if sys.argv[0] == 'service.pushbullet.com' and len(sys.argv) < 2:
		import main
		main.main()
	else:
		try:
			args = None
			if len(sys.argv) > 1:
				args = sys.argv[1:]
			
			if args:
				import main
				main.handleArg(args[0])
			else:
				PushbulletService()
		except:
			import traceback
			traceback.print_exc()


lib/maps.py
import xbmc, urllib, os, time, sys #@UnresolvedImport
class Maps:
	def __init__(self):
		self.setMapSource()
		self.setMapType()
		
		self.zoom =  {	'country':2,
						'region':4,
						'locality':9,
						'neighborhood':13,
						'photo':15}
						
	def setMapSource(self,source='google'):
		self.map_source = source
		if self.map_source == 'yahoo':
			import elementtree.ElementTree as et #@UnresolvedImport
			self.ET = et
		#self.zoom =  {	'country':int(__settings__.getSetting('country_zoom')),
		#				'region':int(__settings__.getSetting('region_zoom')),
		#				'locality':int(__settings__.getSetting('locality_zoom')),
		#				'neighborhood':int(__settings__.getSetting('neighborhood_zoom')),
		#				'photo':int(__settings__.getSetting('photo_zoom'))}
		#self.default_map_type = ['hybrid','satellite','terrain','roadmap'][int(__settings__.getSetting('default_map_type'))]
	
	def setMapType(self,mtype='hybrid'):
		self.default_map_type = mtype
		
	def getMap(self,lat,lon,zoom=15,width=640,height=360,scale=2,marker=False,return_url_only=False):
		#640x36
		source = self.map_source
		lat = str(lat)
		lon = str(lon)
		zoom = str(self.zoom.get(zoom,zoom))
		#create map file name from lat,lon,zoom and time. Take that thumbnail cache!!! :)
		fnamebase = (lat+lon+zoom+str(int(time.time()))).replace('.','')
		ipath = os.path.join(CACHE_PATH,fnamebase+'.jpg')
		mark = ''
		if lon == 'None':
			address = lat
		else:
			address = lat+","+lon
			
		if marker:
			if source == 'osm': 
				mark = '&mlat0=' + lat + '&mlon0=' + lon + '&mico0=0'
			elif source == 'yahoo':
				mark = ''
			else:
				mark = '&markers=color:blue|' + address
		if source == 'osm':
			url = "http://ojw.dev.openstreetmap.org/StaticMap/?lat="+lat+"&lon="+lon+"&z="+zoom+"&w="+str(width)+"&h="+str(height)+"&show=1&fmt=jpg"
		elif source == 'yahoo':
			#zoom = str((int((21 - int(zoom)) * (12/21.0)) or 1) + 1)
			zoom = self.translateZoomToYahoo(zoom)
			xml = urllib.urlopen("http://local.yahooapis.com/MapsService/V1/mapImage?appid=BteTjhnV34E7M.r_gjDLCI33rmG0FL7TFPCMF7LHEleA_iKm6S_rEjpCmns-&latitude="+lat+"&longitude="+lon+"&image_height="+str(height)+"&image_width="+str(width)+"&zoom="+zoom).read()
			url = self.ET.fromstring(xml).text.strip()
			url = urllib.unquote_plus(url)
			if 'error' in url: return ''
		else:
			url = "http://maps.google.com/maps/api/staticmap?center="+address+"&zoom="+zoom+"&size="+str(width)+"x"+str(height)+"&sensor=false&maptype="+self.default_map_type+"&scale="+str(scale)+"&format=jpg"

		if return_url_only:
			return url + mark

		fname,ignore  = urllib.urlretrieve(url + mark,ipath)
		return fname

	def translateZoomToYahoo(self,zoom):
		#Yahoo and your infernal static maps 12 level zoom!
		#This matches as closely as possible the defaults for google and osm while allowing all 12 values
		zoom = 16 - int(zoom)
		if zoom < 1: zoom = 1
		if zoom >12: zoom = 12
		return str(zoom)
		
	def doMap(self):
		clearDirFiles(CACHE_PATH)
		params = self.getParams()
		self.setMapSource(params.get('source','google'))
		self.setMapType(params.get('type','hybrid'))
		
		image = self.getMap(sys.argv[2],sys.argv[3],params.get('zoom','photo'),width=640,height=360,scale=2,marker=True)
		xbmc.executebuiltin('SlideShow('+CACHE_PATH+')')
		
	def getParams(self):
		params=sys.argv[4]
		param={}
		if len(params) < 2:
			self._params = param
			return {}
		cleanedparams=params.replace('?','')
		if params.endswith('/'): params=params[0:len(params)-2]
		pairsofparams=cleanedparams.split('&')
		for i in range(len(pairsofparams)):
			splitparams={}
			splitparams=pairsofparams[i].split('=')
			if (len(splitparams))==2:
				param[splitparams[0]]=splitparams[1]	
		return param
		
def clearDirFiles(filepath):
	if not os.path.exists(filepath): return
	for f in os.listdir(filepath):
		f = os.path.join(filepath,f)
		if os.path.isfile(f): os.remove(f)

CACHE_PATH = xbmc.translatePath('special://profile/addon_data/%s/maps/' % 'service.pushbullet.com')
if not os.path.exists(CACHE_PATH): os.makedirs(CACHE_PATH)
if __name__ == '__main__':
	Maps().doMap()

lib/devices.py
# -*- coding: utf-8 -*-
import Queue
import PushbulletTargets
import util
import pushhandler

class KodiDevice(PushbulletTargets.Device):
	def init(self):
		self.queue = Queue.Queue()

	def clear(self):
		push = True
		while push: push = self.getNext()

	def hasPush(self):
		return not self.queue.empty()

	def getNext(self):
		if self.queue.empty(): return None
		try:
			data = self.queue.get_nowait()
			self.queue.task_done()
			return data
		except Queue.Empty:
			pass
		return None

	def link(self,data):
		if not pushhandler.canHandle(data):
			pushhandler.checkForWindow()
			return

		self.queue.put_nowait(data)
		return False

	def file(self,data):
		if not pushhandler.canHandle(data):
			pushhandler.checkForWindow()
			return

		self.queue.put_nowait(data)
		return False

	def note(self,data):
		if not pushhandler.canHandle(data):
			pushhandler.checkForWindow()
			return

		self.queue.put_nowait(data)
		return False

	def list(self,data):
		if not pushhandler.canHandle(data):
			pushhandler.checkForWindow()
			return

		self.queue.put_nowait(data)
		return False
	
	def address(self,data):
		if not pushhandler.canHandle(data):
			pushhandler.checkForWindow()
			return

		self.queue.put_nowait(data)
		return False

def getDefaultKodiDevice(ID,name):
	assert ID and name, 'Must provide device ID and name'
	return KodiDevice(ID,name)

lib/gui.py
# -*- coding: utf-8 -*-
import xbmc, xbmcgui, time, urllib, os
import util
import pushhandler
import PushbulletTargets
import devices
import maps
import YDStreamExtractor as StreamExtractor 

T = util.T

CACHE_PATH = os.path.join(xbmc.translatePath(util.ADDON.getAddonInfo('profile')),'cache')
if not os.path.exists(CACHE_PATH): os.makedirs(CACHE_PATH)

def getCachedData(ID):
	path = os.path.join(CACHE_PATH,ID)
	if os.path.exists(path):
		with open(path,'r') as f: return f.read()
	return None

def cacheData(ID,data):
	path = os.path.join(CACHE_PATH,ID)
	with open(path,'w') as f: f.write(data)

def deleteCachedData(ID):
	path = os.path.join(CACHE_PATH,ID)
	if os.path.exists(path): os.remove(path)

def cleanCache(used_ids):
	for ID in os.listdir(CACHE_PATH):
		if not ID in used_ids: deleteCachedData(ID)

def selectDevice(client,extra=T(32050)):
	deviceMap = {}
	try:
		for d in client.getDevicesList():
			if not d.get('active'): continue
			deviceMap[d['iden']] = d['nickname']
	except PushbulletTargets.PushbulletException, e:
		xbmcgui.Dialog().ok(T(32051).upper(), '{0}:'.format(T(32051)),'',e.message)
		return
	
	idx = xbmcgui.Dialog().select(T(32052),deviceMap.values() + [extra])
	if idx < 0: return
	if idx >= len(deviceMap.keys()): return ''
	return deviceMap.keys()[idx]

class BaseWindow(xbmcgui.WindowXML):
	def __init__(self,*args,**kwargs):
		self._closing = False
		self._winID = ''

	def onInit(self):
		self._winID = xbmcgui.getCurrentWindowId()
		
	def setProperty(self,key,value):
		if self._closing: return
		xbmcgui.Window(self._winID).setProperty(key,value)
		xbmcgui.WindowXMLDialog.setProperty(self,key,value)
		
	def doClose(self):
		self._closing = True
		self.close()

class PushbulletWindow(BaseWindow):
	def __init__(self,*args,**kwargs):
		self.client = PushbulletTargets.Client(util.getToken())
		self.lastSelected = None
		self.viewMode = 'SELF'
		self.initViewMode()
		self.pushes = []
		BaseWindow.__init__(self,*args,**kwargs)

	def initViewMode(self):
		setID = util.getSetting('selected_device')
		if setID:
			self.viewMode = setID
		else:
			defModeIdx = util.getSetting('default_view_mode',0)
			viewMode = ['SELF','ALL','LAST'][defModeIdx]

			if viewMode == 'LAST':
				last = util.getSetting('last_view_mode')
				self.viewMode = last or 'SELF'
			else:
				self.viewMode = viewMode

	def onInit(self):
		BaseWindow.onInit(self)
		self.setProperty('loading','1')
		self._winID = xbmcgui.getCurrentWindowId()
		self.pushList = self.getControl(101)
		token = util.getSetting('token')
		if not token: return
		
		loadVideoThumbs = util.getSetting('load_video_thumbs',False)
		kodiDevice = devices.getDefaultKodiDevice(util.getSetting('device_iden'),util.getSetting('device_name'))
		if not kodiDevice: return
		self.pushes = []
		pushes = self.client.pushes()
		if not pushes: return
		items = []
		cacheIDs = []
		self.pushes = []

		for p in pushes: #Keep all IDs cached so that we don't cause a delay when changing view
			if p.get('active'):
				cacheIDs.append(p.get('iden'))

		if self.viewMode == 'SELF':
			self.pushes = [p for p in pushes if p.get('active') and p.get('target_device_iden') == kodiDevice.ID]
		elif self.viewMode == 'ALL':
			self.pushes = [p for p in pushes if p.get('active')]
		elif self.viewMode:
			self.pushes = [p for p in pushes if p.get('active') and p.get('target_device_iden') == self.viewMode]

		for push in self.pushes:
			iden = push.get('iden')

			title = push.get('title',push.get('name',push.get('file_name','')))
			bg = push.get('image_url','')
			info = push.get('url','')
			mediaIcon = ''
			media = ''

			if push.get('type') == 'address':
				bg = maps.Maps().getMap(urllib.quote(push.get('address','')),'None',marker=True,return_url_only=True)
			elif push.get('type') == 'link':
				url = push.get('url')
				if StreamExtractor.mightHaveVideo(url):
					media = 'video'
					if loadVideoThumbs:
						bg = getCachedData(iden)
						if not bg:
							bg = StreamExtractor.getVideoInfo(url).thumbnail
							cacheData(iden,bg)
				else:
					media = pushhandler.getURLMediaType(url)
				if not title:
					title = url.rsplit('/',1)[-1]
			elif push.get('type') == 'file':
				info = urllib.unquote(push.get('file_url',''))
				if push.get('file_type','').startswith('image/'):
					media = 'image'
				elif push.get('file_type','').startswith('audio/'):
					media = 'music'
				elif push.get('file_type','').startswith('video/'):
					media = 'video'
			if media:
				mediaIcon = 'service-pushbullet-com-icon_{0}.png'.format(media)

			item = xbmcgui.ListItem(title,iconImage='service-pushbullet-com-{0}.png'.format(push.get('type','')))

			desc = push.get('body',push.get('address',''))
			if push.get('type') == 'list':
				li = []
				ct = 0
				for i in push.get('items',[]):
					li.append(i.get('text',''))
					ct+=1
					if ct > 50: break
				desc = ', '.join(li)
			desc = '[CR]'.join(desc.splitlines()[:4])
			item.setProperty('description',desc)
			item.setProperty('info',info)
			item.setProperty('sender', push.get('sender_email',''))
			item.setProperty('media_icon',mediaIcon)
			item.setProperty('background',bg)
			#item.setProperty('date',time.strftime('%m-%d-%Y %H:%M',time.localtime(push.get('created',0))))
			item.setProperty('date','{0} {1}'.format(util.durationToShortText(time.time() - push.get('created',0)),T(32053)))
			items.append(item)

		self.setProperty('loading','0')
		self.pushList.reset()
		self.pushList.addItems(items)

		if items: self.setFocusId(101)
		self.reSelect()
		cleanCache(cacheIDs)

	def onClick(self,controlID):
		if controlID == 101:
			push = self.getSelectedPush()
			if not push: return
			if not pushhandler.handlePush(push,from_gui=True):
				xbmcgui.Dialog().ok(T(32054),'{0}:'.format(T(32054)),'',T(32055))

	def onAction(self,action):
		try:
			if action.getId() == 11: #xbmcgui.ACTION_SHOW_INFO:
				self.onInit()
			elif action.getId() == 117: #xbmcgui.ACTION_CONTEXT_MENU:
				self.doMenu()
			push = self.getSelectedPush()
			if push: self.lastSelected = push.get('iden')
		except:
			import traceback
			xbmc.log(traceback.format_exc())
		finally:
			BaseWindow.onAction(self,action)
		
	def getSelectedPush(self):
		selected = self.pushList.getSelectedPosition()
		if selected < 0 or selected >= len(self.pushes): return
		return self.pushes[selected]

	def reSelect(self):
		if not self.lastSelected: return
		idx = next((i for i in range(len(self.pushes)) if self.pushes[i].get('iden') == self.lastSelected),-1)
		if idx < 0:
			self.lastSelected = None
		else:
			self.getControl(101).selectItem(idx)

	def doMenu(self):
		selected = self.pushList.getSelectedPosition()
		options = []
		if self.viewMode != 'ALL':
			options.append(('show_all',T(32056)))
		if self.viewMode != 'SELF':
			options.append(('show_self',T(32057)))
		options.append(('show_device',T(32058)))
		if selected >= 0:
			push = self.pushes[selected]
			if push.get('type') in ('file','link'):
				options.append(('download',T(32059)))
			options.append(('delete',T(32060)))
		idx = xbmcgui.Dialog().select(T(32061),[o[1] for o in options])
		if idx < 0: return
		choice = options[idx][0]
		
		if choice == 'download':
			import os
			d = util.Downloader()
			targetDir = os.path.join(xbmc.translatePath(util.ADDON.getAddonInfo('profile')),'temp')
			if not os.path.exists(targetDir): os.makedirs(targetDir)
			finalTargetDir = d.chooseDirectory()
			if not finalTargetDir: return
			
			url = push.get('url')
			if push.get('type') == 'link':
				vid = StreamExtractor.getVideoInfo(url)
				if vid:
					return d.youtubeDLDownload(vid,targetDir,finalTargetDir)
			elif push.get('type') == 'file':
				url = push.get('file_url')

			fname, ftype = d.downloadURL(targetDir,url,fname=push.get('file_name'),final_target_dir=finalTargetDir)
			if fname:
				xbmcgui.Dialog().ok(T(32062),'{0}:'.format(T(32063)),'[CR]',fname)
			else:
				xbmcgui.Dialog().ok(T(32064),'[CR]',T(32064))

		elif choice == 'delete':
			#set last selected to the item above, or the item below if we are at the top
			closest = selected - 1
			if closest < 0:
				self.lastSelected = None
			else:
				self.lastSelected = self.pushes[closest].get('iden')

			self.client.deletePush(push)
			self.onInit()
		elif choice == 'show_all':
			self.viewMode = 'ALL'
			self.onInit()
		elif choice == 'show_self':
			self.viewMode = 'SELF'
			self.onInit()
		elif choice == 'show_device':
			deviceID = selectDevice(self.client)
			if not deviceID: return
			self.viewMode = deviceID
			self.onInit()
		util.setSetting('last_view_mode',self.viewMode)

class ImageViewWindow(BaseWindow):
	def __init__(self,*args,**kwargs):
		self.url = kwargs.get('url','')
		BaseWindow.__init__(self,*args,**kwargs)

	def onInit(self):
		BaseWindow.onInit(self)
		self.setProperty('image_url',self.url)

class NoteViewWindow(BaseWindow):
	def __init__(self,*args,**kwargs):
		self.text = kwargs.get('text','')
		BaseWindow.__init__(self,*args,**kwargs)

	def onInit(self):
		BaseWindow.onInit(self)
		self.getControl(100).setText(self.text)
		import xbmc
		xbmc.sleep(100) #Attempt to give scrollbar time to becomve visible so we can focus it
		if xbmc.getCondVisibility('Control.IsVisible(101)'): self.setFocusId(101) #Prevent log message by checking visibility first

class ListViewWindow(BaseWindow):
	def __init__(self,*args,**kwargs):
		self.data = kwargs.get('data','')
		self.client = PushbulletTargets.Client(util.getToken())
		BaseWindow.__init__(self,*args,**kwargs)

	def onInit(self):
		BaseWindow.onInit(self)
		self.listControl = self.getControl(101)
		items = []
		for i in self.data.get('items',[]):
			item = xbmcgui.ListItem(i.get('text'))
			item.setProperty('checked',i.get('checked') and '1' or '') 
			items.append(item)
		self.listControl.reset()
		self.listControl.addItems(items)
		if items: self.setFocusId(101)

	def refresh(self):
		idx = self.listControl.getSelectedPosition()
		for p in self.client.pushes():
			if p.get('iden') == self.data.get('iden'):
				self.data = p
				break

		self.onInit()
		if idx < 0 or idx >= self.listControl.size(): return
		self.listControl.selectItem(idx)

	def onAction(self,action):
		try:
			if action.getId() == 11: #xbmcgui.ACTION_SHOW_INFO:
				self.refresh()
		finally:
			BaseWindow.onAction(self,action)

	def onClick(self,controlID):
		if controlID == 101:
			idx = self.listControl.getSelectedPosition()
			if idx < 0: return
			item = self.listControl.getListItem(idx)
			checked = not self.data['items'][idx].get('checked')
			self.data['items'][idx]['checked'] = checked
			item.setProperty('checked',checked and '1' or '')
			self.client.modifyPush(self.data)

WINDOW_FONTS = {
	'default':{
		'@font10@':'font10', #Small
		'@font10T@':'font10', #Small Title
		'@font13@':'font13', #Normal
		'@font13T@':'font13' #Normal Title
	}
}

try:
	import json
except:
	import simplejson as json
	
SKIN_XML_PATH = os.path.join(xbmc.translatePath(util.ADDON.getAddonInfo('path')),'resources','skins','Main','720p')

with open(os.path.join(SKIN_XML_PATH,'fonts.json'),'r') as j:
	WINDOW_FONTS.update(json.load(j))
		
def openWindow(winClass,winXML,**kwargs):
	winXMLPath = os.path.join(SKIN_XML_PATH,winXML)
	workingXMLPath = os.path.join(SKIN_XML_PATH,'service.pushbullet.com-working.xml')

	skinName = util.skinName()
	fonts = WINDOW_FONTS['default']
	if skinName in WINDOW_FONTS: fonts = WINDOW_FONTS[skinName]

	with open(winXMLPath,'r') as infile:
		with open(workingXMLPath,'w') as outfile:
			out = infile.read()
			for f,r in fonts.items():
				out = out.replace(f,r)
			outfile.write(out)
	w = winClass('service.pushbullet.com-working.xml',util.ADDON.getAddonInfo('path'),'Main','720p',**kwargs)
	w.doModal()
	del w
	os.remove(workingXMLPath)

def showImage(url):
	openWindow(ImageViewWindow,'service.pushbullet.com-image.xml',url=url)

def showNote(text):
	openWindow(NoteViewWindow,'service.pushbullet.com-note.xml',text=text)

def showList(data):
	openWindow(ListViewWindow,'service.pushbullet.com-list.xml',data=data)

def start():
	openWindow(PushbulletWindow,'service.pushbullet.com-pushes.xml')

		